[WIP] pg_rewind datadirs sync improvement

hughcapet opened the pull request:
**Problem description**
If a leader that is not fast enough in archiving WAL segments (e.g. network issues, high CPU/Disk load...) fails and a replica is successfully promoted, it can be out for long enough time to let the promoted replica rotate the WAL segments that were not archived. As long as archive_mode != always (allow replica to also archive WALs) and replica (by design) doesn't send anything from previous timeline to archive on promote (except for .partial and .history files) - archive will never receive the missing files.
The only place where the WAL segments are still present is the failed leader's FS.

However when we try to revive the leader, we first run pg_rewind utility in order retrieve the missing WAL files from archive or to remove the WALs that could be generated on the old leader after the replica promoted (after the point of divergence). To our surprise, pg_rewind is quite trivial: it basically syncs the content of pg_wal directory of the old leader with the new leader's one. Including deletion of all WALs that are not present on the current leader (promoted replica). After pg_rewind finishes successfully we lost the non-archived WALs.

**Proposed solution**
Make pg_rewind to be a little bit wiser in terms of creating filemap for WALs: preserve WAL segments, which contain those potentially lost records (> the last common checkpoint and  < the point of divergence), on the target.

CyberDem0n commented on src/bin/pg_rewind/filemap.c:
```diff
@@ -207,6 +228,33 @@ lookup_filehash_entry(const char *path)
+	skipwal_t   *entry;
+	bool		found;
+
+	if (keepwalhash)
```
why do we have this check?

hughcapet replied:
in case the insertion function will be accidentally called before the hash initialization

CyberDem0n replied:
Well... There is no similar check in the insert_filehash_entry()

hughcapet replied:
`insert_keepwalhash_entry` can potentially be called from 3 different `SimpleXLogPageRead` invocations (one of them is before the initialization). `insert_filehash_entry` is less likely to fall into this problem. But maybe it is better to add Assert there...

hughcapet commented:
@Stolb27 it potentially decreases the number of `gprecoverseg -F` calls, doesn't it?

darthunix commented on src/bin/pg_rewind/parsexlog.c:
```diff
@@ -297,6 +301,12 @@ SimpleXLogPageRead(XLogReaderState *xlogreader, XLogRecPtr targetPagePtr,
 
+		if (private->keepWalSeg)
+		{
+			snprintf(xlogfpath, MAXPGPATH, XLOGDIR "/%s", xlogfname);
```
Whether do we really need an `XLOGDIR` prefix for the hash key? At first glance, it seems to be redundant here.

hughcapet replied:
a person straggling with the speed of string processing came into the chat :D

it will be then less straightforward to use keepwalhash_entry_exists I guess...

darthunix commented on src/bin/pg_rewind/filemap.c:
```diff
@@ -64,6 +64,27 @@ static file_entry_t *lookup_filehash_entry(const char *path);
+#define SH_DEFINE
+#include "lib/simplehash.h"
+
+static keepwalhash_hash *keepwalhash;
```
You check `keepwalhash` initialization in `insert_keepwalhash_entry()` function, so
```suggestion
static keepwalhash_hash *keepwalhash = NULL;
```

darthunix commented on src/bin/pg_rewind/filemap.c:
```diff
@@ -207,6 +228,33 @@ lookup_filehash_entry(const char *path)
+void
+keepwalhash_init(void)
+{
+	keepwalhash = keepwalhash_create(FILEHASH_INITIAL_SIZE, NULL);
```
Do you think that preallocation of 1000 entries is a good expectation here? May be we can get a smaller estimation relying on some WAL-related GUCs?

hughcapet replied:
good question what is actually the max amount of walsegs we can end up here with
 @CyberDem0n ?

hughcapet replied:
will smth like 
max_wal_size / WalSegSz
be the right amount?

hughcapet replied:
how can GUCs give us the **lower** bound for this...?

hughcapet committed:
Be more picky with WAL segment deletion in pg_rewind

Make pg_rewind to be a bit wiser in terms of creating filemap:
preserve on the target all WAL segments that contain records between the
last common checkpoint and the point of divergence.

Co-authored-by: Alexander Kukushkin <cyberdemn@gmail.com>

hughcapet commented:
old version: https://github.com/hughcapet/postgres/tree/pg_rewind-wal-deletion-old

CyberDem0n commented on src/bin/pg_rewind/parsexlog.c:
```diff
@@ -297,6 +301,16 @@ SimpleXLogPageRead(XLogReaderState *xlogreader, XLogRecPtr targetPagePtr,
+			snprintf(xlogfpath, MAXPGPATH, XLOGDIR "/%s", xlogfname);
+			entry = insert_filehash_entry(xlogfpath);
+
+			entry->action = FILE_ACTION_NONE;
```
We need to fill a bit more fields (similar to process_target_file()):
```c
	entry->target_exists = true;
	entry->target_type = type;
	entry->target_size = size;
	entry->target_link_target = link_target ? pg_strdup(link_target) : NULL;
```

hughcapet replied:
why? it will be anyway done later in process_target_file for these files

CyberDem0n replied:
Hmm... ok.

hughcapet replied:
That’s what I meant - an ugly solution

hughcapet replied:
Well, [the check](https://github.com/hughcapet/postgres/pull/1/files#diff-32d88fe8ad371a4327182459a3d8c97356226bd1a14532f43571da43cd435bc6R272) for action value in process_target_file is actually redundant then (we don's assign target_exists here, it is false there)
